phoenix-load-balancer/src/main/java/org/apache/phoenix/loadbalancer/service/LoadBalancer.java

```diff
@@ -0,0 +1,165 @@
+            cache.start(PathChildrenCache.StartMode.BUILD_INITIAL_CACHE);
+            closeAbles.add(cache);
+            closeAbles.add(curaFramework);
+        }catch(Exception ex){
+            LOG.error("Exception while creating a zookeeper clients and cache",ex);
+            if ((curaFramework != null) && (connectionStateListener != null)){
+                curaFramework.getConnectionStateListenable()
+                        .removeListener(connectionStateListener);
+            }
+            if ((curaFramework != null) && (unhandledErrorListener != null)){
+                curaFramework.getUnhandledErrorListenable()
+                        .removeListener(unhandledErrorListener);
+            }
+            for (Closeable closeable : closeAbles) {
+                CloseableUtils.closeQuietly(closeable);
+            }
+        }
+    }
+
+    /**
+     * Return Singleton Load Balancer every single time.
+     * @return LoadBalancer
+     */
+    public static LoadBalancer getLoadBalancer() {
+        return loadBalancer;
+    }
+
+    /**
+     * return the location of Phoenix Query Server
+     * in form of String. The method return a random PQS server
+     * from the cache.
+     * @return - PQS Location ( host:port)
+     * @throws Exception
+     */
+    public PhoenixQueryServerNode getSingleServiceLocation()  throws Exception{
+        List<PhoenixQueryServerNode> childNodes = conductSanityCheckAndReturn();
+        // get an random connect string
+        int i = ThreadLocalRandom.current().nextInt(0, childNodes.size());
+        return childNodes.get(i);
+    }
+
+    /**
+     * return all the location of Phoenix Query Server from cache
+     * @return - PQS Locations ( host1:port1,host2:port2)
+     * @throws Exception
+     */
+    public List<PhoenixQueryServerNode> getAllServiceLocation()  throws Exception{
+        return conductSanityCheckAndReturn();
+    }
+
+    private List<PhoenixQueryServerNode> conductSanityCheckAndReturn() throws Exception{
+        Preconditions.checkNotNull(curaFramework
+                ," curator framework in not initialized ");
+        Preconditions.checkNotNull(cache," cache value is not initialized");
+        boolean connected = curaFramework.getZookeeperClient().isConnected();
+        if (!connected) {
+            String message = " Zookeeper seems to be down. The data is stale ";
+            Exception exception =
+                    new Exception(message);
+            LOG.error(message, exception);
+            throw exception;
+        }
+        List<String> currentNodes = curaFramework.getChildren().forPath(CONFIG.getParentPath());
+        List<PhoenixQueryServerNode> returnNodes = new ArrayList<>();
+        for(String node:currentNodes) {
+            byte[] bytes = curaFramework.getData().forPath(CONFIG.getParentPath() + "/" + node);
+            returnNodes.add(PhoenixQueryServerNode.fromJsonString(new String(
```

Comment:
Probably want to try/catch around each node? You wouldn't want one bad entry to prevent access to all other good entries.